### PR TITLE
GetLocale API method

### PR DESCRIPTION
**Pull request by Gtonizuka**
Implements a new API endpoint to get wallet locale

## Screenshots/Screencaps

Screenshot 2022-04-25 at 13 44 46

## Manual testing steps

Tested on flask trying to get current locale and then changed it to a different language. (see screenshot)

### CODE REVIEW

**Comment by shanejonas** [comment]
I think we should remove the restricted part for now at least.

File: app/scripts/lib/rpc-method-middleware/handlers/index.js [hunk 1]
```diff
@@ -16,5 +17,6 @@ const handlers = [
   sendMetadata,
   switchEthereumChain,
   watchAsset,
+  wallet_getLanguage
```
Comment (shanejonas) [inline app/scripts/lib/rpc-method-middleware/handlers/index.js:20]: if we want it to only be included in flask we should wrap this line with the:

```
///: BEGIN:ONLY_INCLUDE_IN(flask)
```

File: app/scripts/lib/rpc-method-middleware/handlers/index.js [hunk 2]
```diff
@@ -16,5 +19,8 @@ const handlers = [
   sendMetadata,
   switchEthereumChain,
   watchAsset,
+  ///: BEGIN:ONLY_INCLUDE_IN(flask)
+  walletgetLocale,
```
Comment (FrederikBolding) [inline app/scripts/lib/rpc-method-middleware/handlers/index.js:23]: Perhaps this should be cased like ``walletGetLocale`?

File: shared/constants/permissions.js [hunk 3]
```diff
@@ -5,6 +5,7 @@ export const CaveatTypes = Object.freeze({
 export const RestrictedMethods = Object.freeze({
   eth_accounts: 'eth_accounts',
   ///: BEGIN:ONLY_INCLUDE_IN(flask)
+  wallet_getLocale: 'wallet_getLocale',
```
Comment (FrederikBolding) [inline shared/constants/permissions.js:8]: If it is restricted it needs to be added to https://github.com/MetaMask/metamask-extension/blob/8b9b9af13fabcef4882d4423e35acb4cd3aef4e9/ui/helpers/utils/permission.js

File: app/scripts/lib/rpc-method-middleware/handlers/get-locale.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  try {
+    const walletLocale = await getWalletLocale();
+
+    sendMetrics({
+      event: 'Get locale',
+      category: 'Permission',
+      referrer: {
+        url: origin,
+      },
+    });
```
Comment (brad-decker) [inline app/scripts/lib/rpc-method-middleware/handlers/get-locale.js:49]: [REQ] Unless there are specific props etc we want to track in RPC method handlers, the createRPCMethodTrackingMiddleware file sets up tracking for most RPC methods. If its important to know everytime this method is called (versus a sampling), we would add it to the USER_PROMPTED_EVENT_NAME_MAP in that file ... i don't believe it is necessary to capture every event in this case but I could be wrong.

**Review by brad-decker — DISMISSED** [review]

**Comment by Gtonizuka** [comment]
Moving this back into draft as we need more info from @rekmarks

File: app/scripts/lib/rpc-method-middleware/handlers/get-locale.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+
+async function getLocaleHandler(
+  req,
+  res,
+  _next,
+  end,
+  { getWalletLocale, hasPermission },
+) {
+    if (hasPermission(MESSAGE_TYPE.GET_LOCALE)) {
```
Comment (Gtonizuka) [inline app/scripts/lib/rpc-method-middleware/handlers/get-locale.js:42]: Can you clarify why is this needed @rekmarks ?

File: app/scripts/controllers/permissions/specifications.js [hunk 6]
```diff
@@ -162,6 +162,21 @@ export const getPermissionSpecifications = ({
         }
       },
     },
+    [PermissionKeys['wallet_getLocale']]: {
+      permissionType: PermissionType.RestrictedMethod,
```
Comment (Gtonizuka) [inline app/scripts/controllers/permissions/specifications.js:166]: If `permissionType` is `restrictedMethod` why do we need `if(hasPermission...` in the rpc method implementation (i.e. handlers/get-locale.js below) @rekmarks

File: app/scripts/metamask-controller.js [hunk 7]
```diff
@@ -593,6 +593,7 @@ export default class MetamaskController extends EventEmitter {
           getAllAccounts: this.keyringController.getAccounts.bind(
             this.keyringController,
           ),
+          getWalletLocale: this.getWalletLocale.bind(this),
```
Comment (BelfordZ) [inline app/scripts/metamask-controller.js:596]: Should this be wrapped in `BEGIN:ONLY_INCLUDE_IN(flask)` ?

**Comment by BelfordZ** [comment]
functionally it seems to work great. Nice work. Just left a q about the flask inclusion stuff, but otherwise lgtm!

**Comment by brad-decker** [comment]
@BelfordZ @Gtonizuka do we still want this? If so can we get a rebase and another round of reviews? If not lets close it (for PR queue cleanliness)

**Comment by brad-decker** [comment]
@Gtonizuka I am going to close this, if it ends up being something we want we should reopen and rebase it. Thanks!

**Pull request closed without merging**